manta-trusted-setup/src/groth16/ppot/serialization.rs

```diff
@@ -0,0 +1,964 @@
+        Compressed::Yes => cfg_iter!(powers).for_each(|g| {
+            if !g.is_in_correct_subgroup_assuming_on_curve() {
+                panic!() // This should actually just return the NotInSubgroup error
+            }
+        }),
+    }
+    Ok(powers)
+}
+
+/// Reads `size` many elements of `element_type` from PPoT challenge file.
+#[inline]
+pub fn read_g2_powers<S>(
+    readable_map: &Mmap,
+    element: ElementType,
+    compression: Compressed,
+) -> Result<Vec<GroupAffine<<Parameters as BnParameters>::G2Parameters>>, PointDeserializeError>
+where
+    S: Size,
+{
+    let size = element.num_powers::<S>();
+    let mut powers = Vec::new();
+    let mut start_position = calculate_mmap_position(0, element, compression);
+    let mut end_position = start_position + element.get_size(compression);
+    for _ in 0..size {
+        let mut reader = readable_map
+            .get(start_position..end_position)
+            .expect("cannot read point data from file");
+        if !element.is_g1_type() {
+            let point = match compression {
+                Compressed::No => {
+                    <PpotCeremony as Deserializer<G2Affine, G2>>::deserialize_unchecked(
+                        &mut reader,
+                    )?
+                }
+                Compressed::Yes => {
+                    <PpotCeremony as Deserializer<G2Affine, G2>>::deserialize_compressed(
+                        &mut reader,
+                    )?
+                }
+            };
+            // point will be checked below
+            powers.push(point);
+        } else {
+            panic!("Expected G2 curve points")
+        }
+        start_position = end_position;
+        end_position += element.get_size(compression);
+    }
+    // Do curve point checks in parallel
+    match compression {
+        Compressed::No => cfg_iter!(powers).for_each(|g| curve_point_checks(g).unwrap()),
+        Compressed::Yes => cfg_iter!(powers).for_each(|g| {
+            if !g.is_in_correct_subgroup_assuming_on_curve() {
+                panic!() // This should actually just return the NotInSubgroup error
+            }
+        }),
+    }
+
+    Ok(powers)
+}
+
+/// Reads the proof of correct KZG contribution
+/// This is specific to the compressed PPoT transcript called `response`,
+/// since only it contains this proof.
+#[inline]
+pub fn read_kzg_proof(readable_map: &Mmap) -> Result<Proof<PpotCeremony>, SerializationError> {
+    let position = 64
+        + (PpotCeremony::G1_POWERS + 2 * PpotCeremony::G2_POWERS) * 32
+        + (PpotCeremony::G2_POWERS + 1) * 64;
+    let reader = readable_map
+        .get(position..position + 6 * 64 + 3 * 128)
+        .expect("cannot read point data from file");
+
+    <Proof<_> as CanonicalDeserialize>::deserialize_uncompressed(reader)
+}
+
+/// Extracts a subaccumulator of size `required_powers`. Specific to PPoT challenge file.
```

Comment:
`required_powers` is not anywhere in the function

Reply:
Good catch. I probably originally wrote this function with that as an argument before realizing that the type `C` is what specifies the size of the subaccumulator. I've changed the documentation